src/main/java/fr/sncf/osrd/speedcontroller/generators/ConstructionAllowanceGenerator.java

```diff
@@ -43,10 +50,84 @@ protected double getFirstHighEstimate() {
+        location.updatePosition(schedule.rollingStock.length, position);
+        return location;
+    }
+
+    @Override
+    protected Set<SpeedController> getSpeedControllers(TrainSchedule schedule,
+                                                       double value,
+                                                       double initialPosition,
+                                                       double endPosition) {
+
+        double timestep = 0.01; // TODO: link this timestep to the rest of the simulation
+        var currentSpeedControllers = new HashSet<>(maxSpeedControllers);
+        // running calculation to get the initial speed of the Region Of Interest (ROI)
+        var expectedSpeeds = getExpectedSpeeds(sim, schedule, currentSpeedControllers, timestep,
+                0, initialPosition, schedule.initialSpeed);
+        var initialSpeed = expectedSpeeds.interpolate(initialPosition);
+        double scaleFactor = 1;
+        var targetSpeed = value;
+        // TODO: adapt this to non-constant deceleration
+        var requiredBrakingDistance = Double.max(0,
+                (initialSpeed * initialSpeed - targetSpeed * targetSpeed) / 2 * schedule.rollingStock.timetableGamma);
+        var endBrakingPosition = initialPosition + requiredBrakingDistance;
+        // running calculation starting where the braking ends, to create a scaled MapSpeedController from it
+        var roiSpeeds = getExpectedSpeeds(sim, schedule, currentSpeedControllers, timestep,
+                endBrakingPosition, endPosition, initialSpeed);
+        double maxSpeed = max(roiSpeeds.values());
+        scaleFactor = value / maxSpeed;
+
+        currentSpeedControllers.add(new MapSpeedController(roiSpeeds).scaled(scaleFactor));
+
+        LimitAnnounceSpeedController brakingSpeedController = LimitAnnounceSpeedController.create(
+                initialSpeed,
+                initialSpeed * scaleFactor,
+                endBrakingPosition,
+                schedule.rollingStock.timetableGamma
+        );
+        currentSpeedControllers.add(brakingSpeedController);
+
+        // new running calculation with brakingSpeedController + scaled MapSpeedController
+        // now only the re-acceleration phase is missing
+        var newSpeeds = getExpectedSpeeds(sim, schedule, currentSpeedControllers, timestep,
+                initialPosition, endPosition, initialSpeed);
+
+        // backwards acceleration calculation
+        double speed = roiSpeeds.interpolate(endPosition);
+
+        var location = convertPosition(schedule, sim, endPosition);
+
+        do {
+            var integrator = TrainPhysicsIntegrator.make(timestep, schedule.rollingStock,
+                    speed, location.maxTrainGrade());
+            var action = Action.accelerate(-schedule.rollingStock.getMaxEffort(speed));
+            var update =  integrator.computeUpdate(action, Double.POSITIVE_INFINITY,
+                    -1);
```

Comment:
Do we need a negative acceleration if we also set -1 as directionSign to the update?